src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockImporter.scala

```diff
@@ -0,0 +1,350 @@
+            log.debug(s"block ${Block.number(blocks.head)} import failed")
+            Future.successful((importedBlocks, Some(err)))
+        }
+        .recover {
+          case missingNodeEx: MissingNodeException if syncConfig.redownloadMissingStateNodes =>
+            (importedBlocks, Some(missingNodeEx))
+        }
+    }
+
+  private def importMinedBlock(block: Block, state: ImporterState): Unit =
+    importBlock(block, new MinedBlockImportMessages(block), informFetcherOnFail = false)(state)
+
+  private def importNewBlock(block: Block, peerId: PeerId, state: ImporterState): Unit =
+    importBlock(block, new NewBlockImportMessages(block, peerId), informFetcherOnFail = true)(state)
+
+  private def importBlock(block: Block, importMessages: ImportMessages, informFetcherOnFail: Boolean): ImportFn = {
+    def doLog(entry: ImportMessages.LogEntry): Unit = log.log(entry._1, entry._2)
+
+    importWith({
+      doLog(importMessages.preImport())
+      ledger
+        .importBlock(block)(context.dispatcher)
+        .tap(importMessages.messageForImportResult _ andThen doLog)
+        .tap {
+          case BlockImportedToTop(importedBlocksData) =>
+            val (blocks, receipts) = importedBlocksData.map(data => (data.block, data.td)).unzip
+            broadcastBlocks(blocks, receipts)
+            updateTxAndOmmerPools(importedBlocksData.map(_.block), Seq.empty)
+
+          case BlockEnqueued =>
+            ommersPool ! AddOmmers(block.header)
+
+          case DuplicateBlock => ()
+
+          case UnknownParent => () // This is normal when receiving broadcast blocks
+
+          case ChainReorganised(oldBranch, newBranch, totalDifficulties) =>
+            updateTxAndOmmerPools(newBranch, oldBranch)
+            broadcastBlocks(newBranch, totalDifficulties)
+
+          case BlockImportFailed(error) =>
+            if (informFetcherOnFail) {
+              fetcher ! BlockFetcher.BlockImportFailed(Block.number(block), error)
+            }
+        }
+        .map(_ => Running)
+        .recover {
+          case missingNodeEx: MissingNodeException if syncConfig.redownloadMissingStateNodes =>
+            // state node re-download will be handled when downloading headers
+            doLog(importMessages.missingStateNode(missingNodeEx))
+            Running
+        }
+    })
+  }
+
+  private def broadcastBlocks(blocks: List[Block], totalDifficulties: List[BigInt]): Unit = {
+    val newBlocks = (blocks, totalDifficulties).mapN(NewBlock.apply)
+    broadcastNewBlocks(newBlocks)
+  }
+
+  private def broadcastNewBlocks(blocks: List[NewBlock]): Unit = broadcaster ! BroadcastBlocks(blocks)
+
+  private def updateTxAndOmmerPools(blocksAdded: Seq[Block], blocksRemoved: Seq[Block]): Unit = {
+    blocksRemoved.headOption.foreach(block => ommersPool ! AddOmmers(block.header))
+    blocksRemoved.foreach(block => pendingTransactionsManager ! AddTransactions(block.body.transactionList.toSet))
+
+    blocksAdded.foreach(block => {
```

Comment:
why using `(` and `{` instead of one `{`?

Reply:
But how? I need more than one reference to block, so I can't use `_`. I also have to put `{` and `}` due to 2 statements in this block